### ISSUE TITLE
Question: Performance issues with complex component

### ISSUE THREAD

**Issue by BartVanBeurden** [ractivejs/ractive#2919]
## Description:

I'm currently writing a grid component for ractive at https://github.com/BartVanBeurden/ractive-ez-table   (commit 081f0da4fbe0aec535657a6886fa19266239d10a )
I'm experiencing serious performance issues and am wondering if I'm doing something I shouldn't be doing...

Since it's a fairly large component, I haven't attached any code here.  
It's probably easier to try the demo by cloning the repository and opening demo/index.html in your browser.

Specifically, following interactions are very slow:
- Changing the order of columns by dragging/droppping one of the headings
- Adding a grouped column by dragging a column on the top box
- Removing a grouped column by clicking the X button

Following interactions are noticably slow but not terrible:
- Sorting a column or group
- Selecting one or more rows

## Versions affected:

0.8.9

## Platforms affected:

Mainly firefox, to a lesser extent Chrome (IE currently doesn't work, need to fix drag/drop)

## Code:

Main code files:
- https://github.com/BartVanBeurden/ractive-ez-table/blob/081f0da4fbe0aec535657a6886fa19266239d10a/lib/EzTable.js
- https://github.com/BartVanBeurden/ractive-ez-table/blob/081f0da4fbe0aec535657a6886fa19266239d10a/html/EzTable.html

Demo files:
- https://github.com/BartVanBeurden/ractive-ez-table/blob/081f0da4fbe0aec535657a6886fa19266239d10a/demo/app.js
- https://github.com/BartVanBeurden/ractive-ez-table/blob/081f0da4fbe0aec535657a6886fa19266239d10a/demo/index.html

Basically, the grouping works like this: recursively build a tree structure, then flatten the tree so we end up with a simple array of rows (type: group = a group row, type: item = an item). This takes at most a few milliseconds so this isn't where the performance issues arise.

Beyond that, it is a fairly simple template that just loops over the rows, with only two "special" cases:
1. the `~/template("view", col)` helper creates new partials to support custom templates per column
2. the edit template needs to get `~/items[row.index]` instead of simply `row.item` in order to support two-way binding

Are there any performance do's and don't's that I'm infringing on?

**Comment by ghost** [comment]
Would be nice to have a jsfiddle or a Github page for the demo so everyone (including those who don't have immediate access to Git but still have a browser) can check it out. 😄

**Comment by BartVanBeurden** [comment]
Couldn't get a jsfiddle working properly when I posted this question. Anyway:  
Demo: https://bartvanbeurden.github.io/ractive-ez-table/  
Code: https://github.com/BartVanBeurden/ractive-ez-table/tree/081f0da4fbe0aec535657a6886fa19266239d10a

You can notice serious lag when e.g. dragging the "age" column next to the "active" grouping. If you then drag the "age" column on top of the "active" grouping to swap the group order, the lag is even worse.

**Comment by ghost** [comment]
Haven't looked into the code yet, but here's the profiler results. 

![screen shot 2017-03-27 at 2 53 23 pm](https://cloud.githubusercontent.com/assets/1582932/24372812/34933e60-12fd-11e7-908f-83291a8c4f91.png)

![screen shot 2017-03-27 at 2 55 30 pm](https://cloud.githubusercontent.com/assets/1582932/24372896/8328282e-12fd-11e7-9671-aa2e0260e81d.png)

**Comment by sarendipitee** [comment]
@Bart you can mirror your master branch to gh-pages on your repo then
GitHub will serve your demo files so everyone can check out your demo.
That's how I do it in ractive-datatable and others.

**Comment by evs-chris** [comment]
I think template flexibility is biting a bit, and not having a good way to batch operations/move stuff around in one swing is another significant issue.

So, first up, you can actually render a partial from your data. Changing you `{{>~/template('type', var)}}` calls to `{{>column.template}}` or something like that would probably reduce the amount of stuff that goes through the update grinder, and it also eliminates a ton of expression models. I actually added that specifically for this type of scenario where you want a template to be easily user customizable from a sort of meta-data structure. Basically, you can throw your column template into the column definition like `{ label: titleize(name), template: Ractive.parse(`{{${name}}}`), ... }` and still let the user override it. 0.9 takes it a step further and lets you `yield` inlines so that event handlers and references play out the way you'd expect.

Second, two or more splices in a row _may_ be better served by a `merge`, because it would eliminate renders between splice ops. I say _may_ because there are certain scenarios wherein `merge` causes unwanted side-effects, but it's probably fine for this particular situation. This is definitely an area where ractive could use a little improvement.

**Comment by ghost** [comment]
I also notice a lot of function calls on the template. Wondering if computed properties could help in those situations. Pre-shaping the data to a structure workable by the template might just help instead of calling functions on the fly in the template. This way, it eases up load on the rendering, and put all the processing in JS which you have more wiggle room to optimize.

That's what I often do with APIs that return nested data. I tend to flatten them out to a 1-D array using computed props since most of the time... I'm just rendering lists and tables

**Comment by BartVanBeurden** [comment]
I've tried many of these suggestions but it doesn't seem to help much, if at all.

I've also noticed in the DOM inspector that when moving a group, rather than moving the `<tr>` elements around, it keeps the `<tr>` elements in place and changes the contents of the `<td>` elements. So instead of simply reinserting the 500 `<tr>` rows in the correct order, we're changing 500 tr `[data-index]` attributes and updating 500 * 9 `<td>` cells. In a best case scenario where the contents of a `<td>` are a single text node (rather than custom templates), we're looking at 4500 dom updates. In the demo it's probably around 6000 node updates.

If I were to do the dom manipulations myself, I could get away with just 500 `<tr>` insertions. I'm guessing this would solve much of the perceived lag?

But that goes against the idea of a "reactive" ui...

**Comment by drgnorq** [comment]
Hi Bart, have you considered a ‘virtual rendering’ mode?

I’ve made a list component (could be used for grids as well) where each row has a fixed height.
It works drawing only the rows that actually fit into the viewport. Then you add a dummy row on top and a dummy row on the bottom, with an inline css height set to the number of rows that are in front of the rows currently in viewport, multiplied by the height of one row (in pixels). Same for the dummy bottom row (rowheight * number of rows that follow the rows currently in viewport).

This way you render only 50-100 rows at max while the grid can contain 100.000+ rows.

Best regards,
Bart Pype

**Comment by BartVanBeurden** [comment]
First trials seem to be encouraging, unfortunately the width of the columns constantly changes. I guess I could make those fixed too, but it's less than ideal...

**Comment by evs-chris** [comment]
Another thing that might help would be to create a row partial based on the column definitions. Then when the columns chance, there's only one iteration that needs to update rather than three nested iterations.
